### PR TITLE
Implement 'podman run --blkio-weight-device'

### PR DESCRIPTION
**Pull request by hshiina**
Signed-off-by: Hironori Shiina <shiina.hironori@jp.fujitsu.com>

#### What this PR does / why we need it:

`--blkio-weight-device` is not fully implemented and this causes an
unexpected panic when specified because an entry is put into an
uninitialized map at parsing.

This fix implements the `--blkio-weight-device` and adds a system test.

#### How to verify it

Run an added system test.

#### Which issue(s) this PR fixes:
None

### CODE REVIEW

File: pkg/specgen/generate/container.go [hunk 1]
```diff
@@ -272,6 +272,28 @@ func CompleteSpec(ctx context.Context, r *libpod.Runtime, s *specgen.SpecGenerat
 	return warnings, nil
 }
 
+// FinishDeviceWeights takes the temporary representation of the device
```
Comment (mheon) [inline pkg/specgen/generate/container.go:275]: Why not include this in `parseWeightDevices`?

**Comment by edsantiago** [comment]
/hold

This is broken; I need time to submit a proper review and fix. (Nice work, though. The bug is subtle and not a reflection on you. Thank you for this PR!)

**Comment by edsantiago** [comment]
Suggested [diff](https://github.com/containers/podman/files/7546954/12307.diff.txt)

Explanation: the crucial part is the change from `is $output` to `is $lines[1]`. This reveals that `$lomajmin` was empty. It was empty because `lsblk` was outputing `loop0`, not `/dev/loop0`. In the original `is $output` check, the `.*` was happily matching an empty string.

Solution: skip `lsblk` altogether, since `losetup` gives us what we need. And, add some sanity tests to confirm that we don't end up with empty `lodevice` or `lomajmin`.

Again, thank you for your PR! This is very nice work, and the error was a tricky one.

File: pkg/specgen/generate/container.go (hunk 1 above)
Comment (hshiina) [reply pkg/specgen/generate/container.go:275]: I just imitated `FinishThrottleDevices()` without thought.

As I see the history now, `FinishThrottleDevices()` is called in `MakeContainer()` because throttle devices can be configured from an infra container configuration when a pod in created with `--device-read-bps` (#11686).
For a case similar where `--blkio-weight-device` is introduced for a pod in the future, it might be better to implement like this.

**Comment by hshiina** [comment]
Thank you for the suggestion. I will update the PR.

File: pkg/specgen/generate/container.go (hunk 1 above)
Comment (hshiina) [reply pkg/specgen/generate/container.go:275]: While throttle values are absolute bps, device weights are relative. The weights cannot be set from an infra container like throttle values. So, it would be better to set device weights at parsing as suggested.

Comment (hshiina) [reply pkg/specgen/generate/container.go:278]: Sorry for changing again and again. Finally I got this function called in `CompleteSpec()` because `parseWeightDevices` can be used in a remote client.

**Comment by edsantiago** [comment]
@hshiina could you rebase on `main` and resubmit? We had a total meltdown of CI. The latest `main` includes a solution. I am sorry for the inconvenience.

**Comment by edsantiago** [comment]
@hshiina we had another problem causing CI failures. Could you rebase one more time please? I'm sorry for the hassle. (And BTW we have yet another major CI problem right now, so your tests _will_ fail, but the current CI problem can be dealt with using Re-run. I will watch this PR and press Re-run as needed).

**Comment by edsantiago** [comment]
LGTM. @mheon could you take a fresh look and verify that your concerns are addressed? @containers/podman-maintainers PTAL also.

File: pkg/specgen/generate/container.go (hunk 1 above)
Comment (mheon) [reply pkg/specgen/generate/container.go:282]: Basically, I would prefer not to have something modify the SpecGenerator after it has been passed to the server; ideally we'd do it as part of creating the SpecGenerator, but it sounds like that would break the remote client. Instead, can we do this as part of `MakeContainer` and directly modify the OCI spec to be returned, not the SpecGenerator itself?

**Comment by umohnani8** [comment]
LGTM
@mheon PTAL

File: libpod/options.go [hunk 2]
```diff
@@ -1831,6 +1832,31 @@ func WithHostDevice(dev []specs.LinuxDevice) CtrCreateOption {
 	}
 }
 
+func WithDeviceWeights(weights map[string]specs.LinuxWeightDevice) CtrCreateOption {
```
Comment (mheon) [inline libpod/options.go:1835]: This isn't necessary, you can directly alter the OCI spec in https://github.com/containers/podman/blob/main/pkg/specgen/generate/oci.go#L177 which returns the final spec that is used by Libpod. Moving this code to there will remove the need for Libpod changes.

**Comment by rhatdan** [comment]
Could you rebase to current main branch.  There has been a fix for the CI system.

git pull origin main
git rebase -i origin

**Comment by edsantiago** [comment]
CI is green. @containers/podman-maintainers PTAL. @hshiina thanks for your perseverance.
